mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h

```diff
@@ -0,0 +1,421 @@
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value at(unsigned row, unsigned column) const {
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value &operator()(unsigned row, unsigned column) { return at(row, column); }
+  Value operator()(unsigned row, unsigned column) const {
+    return at(row, column);
+  }
+
+  /// Swap \p column with \p otherColumn.
+  void swapColumns(unsigned column, unsigned otherColumn);
+
+  /// Swap \p row with \p otherRow.
+  void swapRows(unsigned row, unsigned otherRow);
+
+  unsigned getNumRows() const { return nRows; }
+
+  unsigned getNumColumns() const { return nColumns; }
+
+  /// Get a copy of the specified \p row.
+  SmallVector<Value> getRow(unsigned row) const;
+
+  /// Set the specified \p row to \p elems.
+  void setRow(unsigned row, ArrayRef<Value> elems);
+
+  /// Fill \p row with the given value \p val.
+  void fillRow(unsigned row, Value val);
+
+  /// Add an extra row at the bottom of the matrix.
+  unsigned appendRow();
+
+  /// Add an extra row at the bottom of the matrix and copy the given elements
+  /// \p elems into the new row.
+  unsigned appendRow(ArrayRef<Value> elems);
+
+  /// Get a copy of the specified \p column.
+  SmallVector<Value> getColumn(unsigned column) const;
+
+  /// Set the specified \p column to \p elems.
+  void setColumn(unsigned col, ArrayRef<Value> elems);
+
+  /// Fill \p col with the given value \p val.
+  void fillColumn(unsigned col, Value val);
+
+  /// Fill the matrix with the given value \p val.
+  void fill(Value val);
+
+  /// Construct a new matrix containing the specified \p rows.
+  MemoryAccessMatrix getRows(std::set<unsigned> rows) const;
+
+  /// Construct a new matrix containing the specified \p columns.
+  MemoryAccessMatrix getColumns(std::set<unsigned> columns) const;
+
+  /// Construct a new matrix containing the sub-matrix specified by \p rows and
+  /// \p columns.
+  MemoryAccessMatrix getSubMatrix(std::set<unsigned> rows,
+                                  std::set<unsigned> columns) const;
+
+  //===----------------------------------------------------------------------===//
+  // Shape Queries
+  //===----------------------------------------------------------------------===//
+
+  /// Returns true if the matrix has equal number of rows and columns.
+  bool isSquare() const;
+
+  /// Returns true if the matrix is the filled with zero values.
+  bool isZero(DataFlowSolver &solver) const;
+
+  /// Returns true if the only non-zero entries are on the diagonal.
+  bool isDiagonal(DataFlowSolver &solver) const;
+
+  /// Returns true if the matrix is the unit matrix.
+  bool isIdentity(DataFlowSolver &solver) const;
+
+  /// Returns true if all non-zero entries are below the diagonal.
```

Comment:
How about the diagonal? The current implementation checks that they are non-zero entries too.

Reply:
ok I get it now, fixed.